header-rewrite-filter/utility.cc

```diff
@@ -59,6 +80,77 @@ bool evaluateExpression(bool operand1, BooleanOperatorType operator_val, bool op
     return operand1 || operand2;
 }
 
+std::tuple<absl::Status, bool> getHeaderValue(Http::RequestOrResponseHeaderMap& headers, std::vector<absl::string_view> arguments, std::function<bool(std::string)> matcher, bool negate) {
+    
+    const Http::LowerCaseString header_key(arguments.at(0));
+    const Envoy::Http::HeaderUtility::GetAllOfHeaderAsStringResult header = Envoy::Http::HeaderUtility::getAllOfHeaderAsString(headers, header_key);
+    absl::string_view values_string_view;
+    
+    try {
+        values_string_view = header.result().value();
+    } catch (std::exception& e) { // header does not exist
+        const bool result = false;
+        const bool apply_negation = negate ? !result : result;
+        return std::make_tuple(absl::OkStatus(), apply_negation);
+    }
+    
+    try {
+        const auto header_vals = StringUtil::splitToken(values_string_view, ",", false, true);
+        const auto num_header_vals = header_vals.size();
+
+        // get header value based on an optional position argument, if no position specified return the last value
+        if (arguments.size() > 2) {
+            return std::make_tuple(absl::InvalidArgumentError("invalid match syntax -- too many arguments to hdr function"), false);
+        }
+
+        int index = std::stoi(std::string(arguments.at(1)));
+
+        if (arguments.size() == 1) { // no position specified
+            index = num_header_vals - 1;
+        } else if (index < 0) { // negative indexing
+            index += num_header_vals;
+        }
+
+        // validate index
+        if ((index < 0) || (index >= num_header_vals)) {
+            return std::make_tuple(absl::InvalidArgumentError("invalid match syntax -- hdr position out of bounds"), false);
+        }
+
+        // get comma-separated value of the header
+        const auto header_val = header_vals.at(index);
+        const std::string source(header_val);
+
+        const bool result = matcher(source);
+        const bool apply_negation = negate ? !result : result;
+        return std::make_tuple(absl::OkStatus(), apply_negation);
+    } catch (std::exception& e) { // should never happen, bounds are checked above
+        return std::make_tuple(absl::InvalidArgumentError("failed to perform boolean match -- " + std::string(e.what())), false);
+    }
+}
+
+std::tuple<absl::Status, bool> getUrlp(Http::RequestOrResponseHeaderMap& headers, std::vector<absl::string_view> arguments, std::function<bool(std::string)> matcher, bool negate) {
+    try {
+        Http::RequestHeaderMap* request_headers = dynamic_cast<Http::RequestHeaderMap*>(&headers); // can fail if invalid config is provided, ie if response tries to get path
+        if (!request_headers) {
```

Comment:
It might be nice to add a check at initialization time too

Reply:
Will make this a separate PR where request/response boolean variables are stored and executed separately.